subsys/net/lib/sockets/sockets_tls.c

```diff
@@ -1930,27 +1949,54 @@ static int ztls_poll_update_ctx(struct tls_context *ctx,
 	const struct fd_op_vtable *vtable;
 	void *obj;
 	int ret;
+	short events = pfd->events;
 
 	obj = z_get_fd_obj_and_vtable(
 		ctx->sock, (const struct fd_op_vtable **)&vtable);
 	if (obj == NULL) {
 		return -EBADF;
 	}
 
+	/* Check if the socket was waiting for the handshake to complete. */
+	if ((pfd->events & ZSOCK_POLLIN) &&
+	    ((*pev)->obj == &ctx->tls_established)) {
+		/* In case handshake is complete, reconfigure the k_poll_event
+		 * to monitor the underlying socket now.
+		 */
+		if ((*pev)->state != K_POLL_STATE_NOT_READY) {
+			ret = z_fdtable_call_ioctl(vtable, obj,
```

Comment:
Initially (before the handshake is complete) the k_poll_event is configured to monitor `ctx->tls_established` semaphore, which indicates whether the handshake is complete or not (so that the poll() won't busy loop if some data is received before the handshake is initiated or is being handled in some other thread). After the semaphore indicates that the handshake is complete, we can monitor the underlying socket for data (as they will get processed by mbedTLS to indicate whether there's application data available or not), so the event is reconfigured.